### PR TITLE
🐛 listDescendents fetch MachinePools

### LINKED ISSUE #2952: listDescendants should also fetch machinepools
**What steps did you take and what happened:**
[A clear and concise description on how to REPRODUCE the bug.]

When deleting a cluster, machinepools are not deleted. This is due to [`listDescendants` not collecting machinepools](https://github.com/kubernetes-sigs/cluster-api/blob/99a95aaa59492924b04e6dddbea17992fecca24d/controllers/cluster_controller.go#L361-L392).

**What did you expect to happen:**
MachinePools to be deleted when a cluster is deleted.

**Environment:**

- Cluster-api version: master

/kind bug

### PR DESCRIPTION
**Pull request by prankul88**
**What this PR does / why we need it**:
This PR makes sure`listDescendents` function fetches MachinePools object too. This enables MachinePools to be deleted as well when cluster is deleted.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2952

### CODE REVIEW

File: controllers/cluster_controller.go [hunk 1]
```diff
@@ -357,10 +359,17 @@ func (c *clusterDescendants) descendantNames() string {
 	if len(workerMachineNames) > 0 {
 		descendants = append(descendants, "Worker machines: "+strings.Join(workerMachineNames, ","))
 	}
+	machinePoolNames := make([]string, len(c.machinePools.Items))
+	for i, machinePool := range c.machinePools.Items {
+		machinePoolNames[i] = machinePool.Name
+	}
+	if len(machinePoolNames) > 0 {
+		descendants = append(descendants, "Machine pools: "+strings.Join(machinePoolNames, ","))
+	}
```
Comment (detiber) [inline controllers/cluster_controller.go:368]: This should likely be wrapped with a check to see if the feature gate for MachinePools is enabled, similar to what we are doing here: https://github.com/kubernetes-sigs/cluster-api/blob/2bb56ed2eb2a91cb59b740cb286992952da3dfed/bootstrap/util/configowner.go#L95-L100

**Comment by vincepri** [comment]
Also needs to add some tests

File: controllers/cluster_controller.go [hunk 2]
```diff
@@ -377,6 +386,10 @@ func (r *ClusterReconciler) listDescendants(ctx context.Context, cluster *cluste
 		return descendants, errors.Wrapf(err, "failed to list MachineSets for cluster %s/%s", cluster.Namespace, cluster.Name)
 	}
 
+	if err := r.Client.List(ctx, &descendants.machinePools, listOptions...); err != nil {
+		return descendants, errors.Wrapf(err, "failed to list MachineSets for the cluster %s/%s", cluster.Namespace, cluster.Name)
+	}
+
```
[the remark this answers is not in the document]
Comment (prankul88) [reply controllers/cluster_controller.go]: Done

**Comment by prankul88** [comment]
@detiber Enabled feature gate as you mentioned. Will be adding tests soon

**Comment by vincepri** [comment]
@prankul88 Do you have time to follow-up with some tests? Should be ok to also open an issue and open a PR later.

**Comment by prankul88** [comment]
> @prankul88 Do you have time to follow-up with some tests? Should be ok to also open an issue and open a PR later.

@vincepri I am a little busy with another patch. Would be nice if we can open an issue for the tests. I can open an issue if you like

**Comment by vincepri** [comment]
@prankul88 Thank you!

/approve
/assign @detiber 
/milestone v0.3.7

File: controllers/cluster_controller.go [hunk 3]
```diff
@@ -417,6 +434,10 @@ func (c clusterDescendants) filterOwnedDescendants(cluster *clusterv1.Cluster) (
 		&c.workerMachines,
 		&c.controlPlaneMachines,
 	}
+	if feature.Gates.Enabled(feature.MachinePool) {
+		lists = append(lists, &c.machinePools)
```
Comment (detiber) [inline controllers/cluster_controller.go:438]: Just realized that we should probably not be appending the machinePools to the end of the list, but rather pre-pending them (or at least inserting them prior to `controlPlaneMachines`.

We can handle that as a followup, though.

**Comment by prankul88** [comment]
@detiber @vincepri I have opened an issue #3294 to address tests for this. Thanks

File: controllers/cluster_controller.go (hunk 3 above)
Comment (vincepri) [reply controllers/cluster_controller.go:438]: Might be worthwhile to do it within this PR, or wait for the next release

Comment (prankul88) [reply controllers/cluster_controller.go:438]: Updating

File: controllers/cluster_controller.go [hunk 4]
```diff
@@ -417,6 +434,12 @@ func (c clusterDescendants) filterOwnedDescendants(cluster *clusterv1.Cluster) (
...
+		copy(lists[4:], lists[3:])
+		lists[3] = &c.machinePools
```
Comment (prankul88) [inline controllers/cluster_controller.go:440]: @detiber @vincepri Please see if this logic solves the issue.

File: controllers/cluster_controller.go [hunk 5]
```diff
@@ -417,6 +434,12 @@ func (c clusterDescendants) filterOwnedDescendants(cluster *clusterv1.Cluster) (
...
+	}
```
Comment (vincepri) [inline controllers/cluster_controller.go:441]:
```suggestion
	if feature.Gates.Enabled(feature.MachinePool) {
		lists = append([]runtime.Object{&c.machinePools}, lists...)
	}
```

This _should_ work to prepend to the list

**Comment by prankul88** [comment]
@vincepri PTAL

File: controllers/cluster_controller.go [hunk 6]
```diff
@@ -377,6 +389,11 @@ func (r *ClusterReconciler) listDescendants(ctx context.Context, cluster *cluste
 		return descendants, errors.Wrapf(err, "failed to list MachineSets for cluster %s/%s", cluster.Namespace, cluster.Name)
 	}
 
+	if feature.Gates.Enabled(feature.MachinePool) {
+		if err := r.Client.List(ctx, &descendants.machinePools, listOptions...); err != nil {
+			return descendants, errors.Wrapf(err, "failed to list MachineSets for the cluster %s/%s", cluster.Namespace, cluster.Name)
```
Comment (fabriziopandini) [inline controllers/cluster_controller.go:394]: nit
```suggestion
			return descendants, errors.Wrapf(err, "failed to list MachinePools for the cluster %s/%s", cluster.Namespace, cluster.Name)
```

**Comment by CecileRobertMichon** [comment]
lgtm besides typo

**Comment by vincepri** [comment]
Fixed the typo by accepting @fabriziopandini's suggestion, to move things along for RC.1 :)

**Pull request merged**